kext/FileNVRAM/FileNVRAM.cpp

```diff
@@ -115,49 +118,97 @@ bool FileNVRAM::start(IOService *provider)
     if(!dict) return false;
     setPropertyTable(dict);
     
-    
     if(bootnvram)
     {
         copyEntryProperties(NULL, bootnvram);
         bootnvram->detachFromParent(root, gIODTPlane);
+        mInitComplete = true;
     }
     
     if(mReadOnly)
     {
         // we assume that the bootloader has done its job
         // i.e. populate /chosen/nvram
-        mInitComplete = true;
-        mSafeToSync = true;
-        registerNVRAM();
+        mInitComplete = true;        mSafeToSync = true;
         return true;
     }
     else
     {
-        IOTimerEventSource* mTimer = IOTimerEventSource::timerEventSource(this, timeoutOccurred);
-        
-        if(mTimer)
+        if(VERSION_MAJOR >10)
         {
-            getWorkLoop()->addEventSource( mTimer);
-            mTimer->setTimeoutMS(50); // callback isn't being setup right, causes a panic
-            mSafeToSync = false;
+            // added in 10.6
+            disk_notifier_start_ = addMatchingNotification(gIOMatchedNotification,
+                                                           serviceMatching("IOMediaBSDClient"),
+                                                           FileNVRAM::DiskFound_callback,
+                                                           this, nullptr, 0);
         }
         else
         {
-            earlyInit = true;
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+            // deprecated in 10.6
+            disk_notifier_start_ = addNotification(gIOMatchedNotification,
+                                                   serviceMatching("IOMediaBSDClient"),
+                                                   ((IOServiceNotificationHandler)&(FileNVRAM::DiskFound_callback)),
+                                                   this, nullptr, 0);
+#pragma clang diagnostic pop
+        }
+        
+        
+        if(disk_notifier_start_ == nullptr)
+        {
+            IOLog("FileNVRAM: disk_notifier_start_ == nullptr\n");
+            return false;
         }
     }
     
-    // We don't have initial nvram data from the bootloader, or we couldn't schedule a
-    // timer to read in the nvram file, so start up immediately.
-    if(earlyInit == true)
-    {
-        mSafeToSync = true;
-        registerNVRAM();
-    }
+    return true;
+}
+
+bool FileNVRAM::DiskFound_callback(void *target, void *ref, IOService *service, IONotifier *notifier)
+{
+    FileNVRAM *self = (FileNVRAM *)target;
     
-    mInitComplete = true;
+    notifier->remove();
+    notifier = NULL;
     
-    return true;
+    if(OSDictionary *matching = nameMatching("Apple_HFS"))
```

Comment:
Recent version of XNU (finally) support more than just hfs+ for the rootfs. For example, zfs can now be used.

Why is this code matching on Apple_HFS?

It also looks like this code never actually verifies that the rootfs was mounted, and just assumes that if one IOMediaBSDClient (HFS) class is found, that the rootfs must have been mounted.

Looking at the ioregistery on my mac (with an encrypted rootfs), I see that the EFI partition shows up as an IOMediaBSDClient even though it has *not* mounted. Yes, the EFI partition probably won't be HFS, but it's entirely possible for other HFS partitions to exist on other driver (or none to exist at all).

So... it looks like there may be a race condition with how this code is implemented

*Note: there is an actual rootfs mounted callback in XNU, however they way Apple implemented it, only one kext can reliably register for it (and I think it's a private API). In any case, it looks like more work needs to be done here to ensure everything always works.